### ISSUE TITLE
Plan for Deep Learning in .NET

### ISSUE THREAD

**Issue by briacht** [dotnet/machinelearning#5918]
# Plan for Deep Learning in .NET

This past year we've been working on our plan for deep learning (DL) in .NET which is outlined in this issue below.

We came up with this plan by looking at recent deep learning trends, staying in line with current Microsoft strategy around deep learning, and gathering a variety of customer evidence (which you can read about in the [Customer evidence](#customer-evidence) section below).

Our goal is to execute this plan by the end of 2022.

## Feedback

We would love to hear your feedback about our plan and whether this will fulfill your requirements for deep learning in .NET!

You can leave your feedback directly on this issue.

## The plan

The main stages of the plan are:

1. Make it easier to consume ONNX models in ML.NET using the ONNX Runtime (RT)
2. Fully support and productionize [TorchSharp](https://github.com/dotnet/TorchSharp) for building neural networks in .NET
3. Build a bridge between TorchSharp and ML.NET

As part of this plan, ML.NET will continue to be more top-level and scenario focused, while TorchSharp will power the Deep Learning training and features in .NET.

As we iterate and work on the plan, we will make updates to this issue, including linking to relevant issues and PRs.

### Stage 1: Make it easier to consume ONNX models in ML.NET with the ONNX RT

In ML.NET, you can already consume pre-trained ONNX models via the [ONNX RT](https://github.com/microsoft/onnxruntime).

However, we have seen many issues from customers around the complexity of ONNX model consumption in ML.NET, particularly when trying to figure out the inputs and outputs of the ONNX model (which are required for consumption in ML.NET).

So, we plan to:

- Expose internal functionality to get the input and output schema from ONNX models ([#5917](https://github.com/dotnet/machinelearning/issues/5917)).
- Add documentation outlining step-by-step how to get the inputs and outputs of any ONNX model and then how to consume that ONNX model in ML.NET. We will try and cover a variety of scenarios ([#25863](https://github.com/dotnet/docs/issues/25863)).
- Partner with the ONNX RT team to bring relevant ONNX RT advancements to ML.NET.

### Stage 2: Fully support and productionize TorchSharp

[TorchSharp](https://github.com/dotnet/TorchSharp) is an API that provides .NET bindings for the [PyTorch](https://pytorch.org/) engine.

TorchSharp will be the API-based Neural Network training and inference in .NET; the custom deep learning training that you can do in PyTorch, you will be able to do in .NET with TorchSharp.

As mentioned in the TorchSharp repo, TorchSharp will retain Python naming conventions. This makes it very easy to port PyTorch to .NET TorchSharp as well as to bring Python examples to .NET.

In order to bring TorchSharp to production, we will:

- Move TorchSharp from the Xamarin repo to the .NET Foundation repo
- Add missing features and optimizers from PyTorch 1.6-1.9
- Update infrastructure for signed builds, tests, and SDLs

Ongoing support work will include:

- Issue and PR management of the repo
- Reacting to new versions of PyTorch (~2x a year)

### Stage 3: Build a bridge between TorchSharp and ML.NET

Once TorchSharp is fully productionized, it will power higher-level ML.NET deep learning APIs which will follow .NET naming conventions and "feel" like .NET.

We will surface the following through ML.NET (powered by TorchSharp):

- Scenario-focused APIs for transfer learning, like the [Image Classification API](https://docs.microsoft.com/dotnet/api/microsoft.ml.vision?view=ml-dotnet); there will be an API for each scenario, like Object Detection and NLP
- Generic transfer learning API for custom scenarios that don't fit the scenario-focused APIs
- Simplified ML.NET APIs and tooling for building neural networks from scratch (this will take a bit more investigation and planning)

Here is a diagram which visually outlines the transfer learning part of the plan:

![ML.NET transfer learning diagram](https://user-images.githubusercontent.com/10437687/131400671-df5e067e-503b-444b-a7a9-d6c68d8dd6bd.png)

### Out of scope

For this plan, it is currently out of scope to port over the PyTorch domain libraries (Torch Vision, Torch Audio, and Torch Text).

## Customer evidence

After talking to a variety of .NET customers, we discovered the following:

- Lack of deep learning support is a top pain point/blocker for using ML.NET
- Customers are forced to turn to Python-based frameworks for training DL models when scenario is not covered by ML.NET (mainly TensorFlow, Keras, and PyTorch)
- Customers have a variety of DL scenarios, the most popular of which include Image Classification, Named Entity Recognition (NER), NLP-based scenarios, and Object Detection
- Almost all customers said they prefer their E2E machine learning workflow, from training to consumption, to be in .NET, and they don’t care about the underlying framework ML.NET uses for training deep learning models
- A majority of customers we talked to want ability to build neural networks from scratch using .NET
- Several customers said adding transfer learning for their scenario would unblock them
- One customer we talked to needed better support for consuming pre-trained models which is also evident in the number of issues filed around consuming ONNX models.
- Customers were split in terms of tooling and automation: some prefer a code-first / API approach when building neural networks from scratch, while some prefer tooling.
- Customers were also split in terms of automation when training deep learning models: some prefer no automation and full control when training, some want full automation, and some want a combination of automation with a lot of control

Additionally, the following issues filed in the repo add further evidence for backing up our deep learning plan:

Image segmentation / regression

- [#5276](https://github.com/dotnet/machinelearning/issues/5276)
- [#5595](https://github.com/dotnet/machinelearning/issues/5595)
- [#5595](https://github.com/dotnet/machinelearning/issues/5595)

NLP

- [#5159](https://github.com/dotnet/machinelearning/issues/5159)
- [#5874](https://github.com/dotnet/machinelearning/issues/5874)
- [#955](https://github.com/dotnet/machinelearning/issues/955)

Local object detection

- [#5863](https://github.com/dotnet/machinelearning/issues/5863)
- [#1091](https://github.com/dotnet/machinelearning/issues/1091)

Torch support

- [#4053](https://github.com/dotnet/machinelearning/issues/4053)
- [#3439](https://github.com/dotnet/machinelearning/issues/3439)

ONNX consumption

- [#5810](https://github.com/dotnet/machinelearning/issues/5810)
- [#5855](https://github.com/dotnet/machinelearning/issues/5855)
- [#5830](https://github.com/dotnet/machinelearning/issues/5830)
- [#5597](https://github.com/dotnet/machinelearning/issues/5597)
- [#5484](https://github.com/dotnet/machinelearning/issues/5484)
- [#5549](https://github.com/dotnet/machinelearning/issues/5549)
- [#4578](https://github.com/dotnet/machinelearning/issues/4578)
- [#4335](https://github.com/dotnet/machinelearning/issues/4335)

Other

- [#5734](https://github.com/dotnet/machinelearning/issues/5734)
- [#5713](https://github.com/dotnet/machinelearning/issues/5713)
- [#5372](https://github.com/dotnet/machinelearning/issues/5372)

**Comment by justinormont** [comment]
> Simplified ML.NET APIs and tooling for building neural networks from scratch (this will take a bit more investigation and planning)

You can take a look at how we exposed CNTK in TLC. Two styles, (1) fully custom designed NNs _[do it myself]_, and (2) a simplification layer which created a network to you specifications _[do it for me]_. 

For 2, you specify the network type you want created, for example a fully-connected (aka feed-forward) NN, and specify the parameters of the network, like number of hidden layers, number of neurons wide, learning rate, training algo, etc. Then we create the NN for you, including your specific input/output sizes (which is a bit difficult to know in TLC / ML․NET before training). 

Exposing common neural network styles in a parameterized form makes it easy for a non-experienced user to create a functional NN, and also lets the AutoML do hyperparameter optimization easily on these networks (e.g. automatically alter the number of layers and learning rate in a fully-connected NN).

**Comment by sps014** [comment]
Great roadmap, i am excited for this.

**Comment by DFMERA** [comment]
Grate roadmap!! Simplifying the consumption of ONNX models is so necessary, I trained an ONNX model once but I could never understand the outpts.
Thank you for adding this to your roadmap!!

**Comment by dsyme** [comment]
This is fantastic!  Great to see the plan coming together :)

**Comment by pjmlp** [comment]
The roadmap looks great! Specially not having to deal with C++ frameworks directly and being able to live in pure .NET workflows.

**Comment by saint4eva** [comment]
> The roadmap looks great! Specially not having to deal with C++ frameworks directly and being able to live in pure .NET workflows.

You are avoiding dealing with C++, but now you are going to deal with python. Performance is going to be lost and the API convention of python will overwhelm .NET and C# developers.

**Comment by dsyme** [comment]
> ...deal with python... performance is going to be lost 

No Python is used in TorchSharp

**Comment by aforoughi1** [comment]
The roadmap should address the Model Explainability, built from bottom-up as a major milestone. 

I found the AutoMl.Net framework significantly improved the rapid modelling process, so I hope to see it as part of the roadmap.

**Comment by pjmlp** [comment]
> You are avoiding dealing with C++, but now you are going to deal with python

ONMX runtime is written in C++.

I am also quite comfortable with C++, including being up to date with C++20, but rather spend my time with languages where productivity comes first.

**Comment by nietras** [comment]
I'm happy to see authoring DL models in C# is finally being prioritized. I wanted to share another idea for this just for context.

Instead of using TorchSharp one could author models in ONNX format via OnnxSharp (https://nietras.com/2021/03/20/introducing-onnxsharp/) if that had an easy to use API for that. With a proper C# API, not the pythonesque TorchSharp API. 

Training could then be done via the ONNX Runtime Training API. Or even TorchSharp if that could load ONNX models.

If I had time I would be working more on fulfilling this idea. 

Having both options would be best of course. Important thing is if ML.NET settles on using ONNX format, there are a lot of options. So happy with that too.

**Comment by kinfey** [comment]
Apple silicon support is important

**Comment by GeorgeS2019** [comment]
@nietras I am aware of and follow your contributions through [OnnxSharp](https://nietras.com/2021/03/20/introducing-onnxsharp/) Keep it up!!!

> Or even TorchSharp if that could load ONNX models.

[TorshSharp save/load ONNX models has been discussed.](https://github.com/dotnet/TorchSharp/issues/328#issue-972933071) We need more people to join in to discuss and support this idea.

**Comment by redradist** [comment]
@briacht
```
As mentioned in the TorchSharp repo, TorchSharp will retain Python naming conventions.
```

I like this one !!

I've asked the same in [`.NET MAUI`](https://github.com/dotnet/maui/issues/522 ) to retain as option the same native API naming conventions, but they did not listen to me ... (

**Comment by redradist** [comment]
@briacht 

Why not to use also https://github.com/SciSharp/TensorFlow.NET ?
It is actively developed and have the same "native" api as TensorFlow ...

Also there are Miguel package https://github.com/migueldeicaza/TensorFlowSharp ... but he does not evolve it unfortunately

**Comment by GeorgeS2019** [comment]
@redradist 
Tensorflow.NET already integrated in ML.NET
TensorflowSharp is no longer as actively developed as Tensorflow.NET

**Comment by redradist** [comment]
@GeorgeS2019
> @redradist
> Tensorflow.NET already integrated in ML.NET

What ?? When ? Where ??
I tried to find it in repo ... do not see it (:

**Comment by GeorgeS2019** [comment]
@redradist  You can get ML.NET book e.g. from Amazon which assist in using Tensorflow.NET in ML.NET

**Comment by redradist** [comment]
@GeorgeS2019
Thanks, but as for me documentation is more important ;)

**Comment by redradist** [comment]
@GeorgeS2019 Seems like I've found it ;)
https://github.com/dotnet/machinelearning/tree/main/src/Microsoft.ML.TensorFlow

**Comment by RoudyES** [comment]
This plan tackles almost every problem that most of us have when it comes to using ML.NET for DL, it is very exciting!
Thank you for listening to your community needs, and, I really hope this plan comes to fruition without many obstacles.

**Comment by saint4eva** [comment]
> > ...deal with python... performance is going to be lost
> 
> No Python is used in TorchSharp

Why python-like apis then?

**Comment by GeorgeS2019** [comment]
@saint4eva We have addressed many of your feedback [here ](https://github.com/dotnet/TorchSharp/issues/308#issuecomment-886159539):-)

**Comment by migueldeicaza** [comment]
> > > ...deal with python... performance is going to be lost
> > 
> > 
> > No Python is used in TorchSharp
> 
> Why python-like apis then?

Those are two separate issues.

TorchSharp is pure C# code calling into the native engine that powers PyTorch.

It also happens to have adopted the shape of APIs of Python, to simplify the job of moving code from Python to C#

**Comment by GeorgeS2019** [comment]
Feedback to 

> TorchSharp adopted the shape of APIs of Python, [to simplify the job of moving code from Python to C#](https://github.com/dotnet/machinelearning/issues/5918#issuecomment-918537115)

using a use case project [ThaiRomanizationSharp](https://github.com/dotnetthailand/ThaiRomanizationSharp/tree/main/src/ThaiRomanizationSharp.Thai2Rom) that depends on the latest TorchSharp
> Like the underlying **TorchSharp library, the C# coding style** often departs from C# standard coding style and descriptive variable names, in order to **match the original Python code**. As improvements are made to the underlying Python code, **it's easier to port them to the C#**.

Additional valuable feedback from the project. Perhaps the TorchSharp and ML.NET teams can comment on that.

> Note that TorchSharp is working hard to support ONNX format, which PyTorch is also working to support. Once both projects have full support, it should be much easier to share models as ONNX, rather than needing the `exportsd` script above.

**Comment by GeorgeS2019** [comment]
**.NET ONNX viewer** to support deep learning porting ONNX model to ML.NET

FYI:  there is a initiative on[ winForm ONNX viewer user control.](https://github.com/fel88/Dendrite/issues/1) **With community support**, this could turn into e.g. project reunion (WinForm/WPF/UWP) control that serves some of the functions of [NETRON  ](https://github.com/lutzroeder/Netron)

![ItKwJv70CZ](https://user-images.githubusercontent.com/49812372/133510059-90d2e1b7-6c10-4e70-b5f4-38df7e330909.gif)

**Comment by GeorgeS2019** [comment]
Do contribute to this [discussion ](https://github.com/dotnet/TorchSharp/discussions/335#discussioncomment-1461360) or this [issue ](https://github.com/dotnet/TorchSharp/issues/403#issuecomment-940467174)if you are interested on TorchSharp ML.NET integration

Currently it is possible in ML.NET to import Tensorflow trained model file .pb
![image](https://user-images.githubusercontent.com/49812372/136858848-332ec873-679b-4dd8-be28-c28dea723776.png)

ML.NET needs native binding to load PyTorch trained model file e.g. pth.  TorchSharp is key to sort out the paths how ML.NET will load such trained file.

The LibTorch is able to load different formats of PyTorch trained model files
- PyTorch `torch.save` model file
- TorchScript ` torch.jit.save` model file

## Question
Shall ML.NET support one of the two or both file formats?

**Comment by Jiuyong** [comment]
> 我很高兴看到在 C# 中编写 Dl 模型终于被优先考虑了。我想分享另一个想法， 只是为了上下文。
> 
> 而不是使用火炬夏普，人们可以通过 OnnxSharp[（https://nietras.com/2021/03/20/introducing-onnxsharp/）](https://nietras.com/2021/03/20/introducing-onnxsharp/)以 ONNX 格式编写模型，如果这有一个容易使用的 API。与适当的 C # Api， 而不是巨蛇火炬夏普 Api 。
> 
> 然后，培训可以通过 ONNX 运行时间培训 API 进行。甚至火炬夏普， 如果这可以加载 Onnx 模型。
> 
> 如果我有时间，我会更加努力地实现这个想法。
> 
> 当然，拥有这两种选择是最好的选择。重要的是，如果 ML.NET 决定使用 ONNX 格式，有很多选项。也很高兴。

I support your idea, but most people will feel that it's too radical. And cutting the ecology of torch is a double-edged sword.

**Comment by Jiuyong** [comment]
> I'm happy to see authoring DL models in C# is finally being prioritized. I wanted to share another idea for this just for context.
> 
> Instead of using TorchSharp one could author models in ONNX format via OnnxSharp (https://nietras.com/2021/03/20/introducing-onnxsharp/) if that had an easy to use API for that. With a proper C# API, not the pythonesque TorchSharp API.
> 
> Training could then be done via the ONNX Runtime Training API. Or even TorchSharp if that could load ONNX models.
> 
> If I had time I would be working more on fulfilling this idea.
> 
> Having both options would be best of course. Important thing is if ML.NET settles on using ONNX format, there are a lot of options. So happy with that too.

I support your idea, but most people will feel that it's too radical. And cutting the ecology of torch is a double-edged sword.

**Comment by saint4eva** [comment]
> > I'm happy to see authoring DL models in C# is finally being prioritized. I wanted to share another idea for this just for context.
> > Instead of using TorchSharp one could author models in ONNX format via OnnxSharp (https://nietras.com/2021/03/20/introducing-onnxsharp/) if that had an easy to use API for that. With a proper C# API, not the pythonesque TorchSharp API.
> > Training could then be done via the ONNX Runtime Training API. Or even TorchSharp if that could load ONNX models.
> > If I had time I would be working more on fulfilling this idea.
> > Having both options would be best of course. Important thing is if ML.NET settles on using ONNX format, there are a lot of options. So happy with that too.
> 
> I support your idea, but most people will feel that it's too radical. And cutting the ecology of torch is a double-edged sword.

Cutting the C# /. NET ecosystem is even a more dangerous double-edged sword. Writing python and forcing C# developers to understand and use python.

They should take a look at what OpenCVSharp is doing. Beautiful apis that conforms to dotnet way of doing things

**Comment by GeorgeS2019** [comment]
> > I'm happy to see authoring DL models in C# is finally being prioritized. I wanted to share another idea for this just for context.
> > Instead of using TorchSharp one could author models in ONNX format via OnnxSharp (https://nietras.com/2021/03/20/introducing-onnxsharp/) if that had an easy to use API for that. With a proper C# API, not the pythonesque TorchSharp API.
> > Training could then be done via the ONNX Runtime Training API. Or even TorchSharp if that could load ONNX models.
> > If I had time I would be working more on fulfilling this idea.
> > Having both options would be best of course. Important thing is if ML.NET settles on using ONNX format, there are a lot of options. So happy with that too.
> 
> I support your idea, but most people will feel that it's too radical. And cutting the ecology of torch is a double-edged sword.

![image](https://user-images.githubusercontent.com/49812372/139819950-7d2db5d9-9407-469e-af35-36a9341f0ec1.png)

@Jiuyong   As you can see, the idea of **OnnxSharp** receives positive support and hardly anyone one here seems to think that it is radical.
   
### Questions
- Where is your source that it is too radical and why?
-  Can you elaborate why double-edged sword? It is an opportunity to learn from your feedback.
- > cutting the ecology of torch is a double-edged sword.

**Comment by GeorgeS2019** [comment]
For users here interested to have some fun hacking out .NET OpenAI Gym ( Reinforcement learning),[ join us FAST ](https://github.com/virtualmlnet/hackathon-2021/issues/6)!!!!

**Comment by JimFFM** [comment]
@chunky  => here is [one possible way](https://github.com/JimFFM/ml-hackathon-2021) to **do Reinforcement Learning using ML.NET**

**Comment by lostmsu** [comment]
This is not doing reinforcement learning using ML.NET. The training happens in Python.

**Comment by chunky** [comment]
@JimFFM I already have a workflow that works well, that makes a round trip through Python [and has a more performant and less byzantine; but also less capable; ecosystem around it]. But I'd really like to skip said round trip and do it directly.

**Comment by torronen** [comment]
"1. Make it easier to consume ONNX models in ML.NET using the ONNX Runtime (RT)" 

Could this fit the use cases for model builder? 
1. Select ONNX model 
2. Model Builder wraps everything in a zip file 
2. Model Builder creates the consumption code and samples

**Comment by WhitWaldo** [comment]
> > > I'm happy to see authoring DL models in C# is finally being prioritized. I wanted to share another idea for this just for context.
> > > Instead of using TorchSharp one could author models in ONNX format via OnnxSharp (https://nietras.com/2021/03/20/introducing-onnxsharp/) if that had an easy to use API for that. With a proper C# API, not the pythonesque TorchSharp API.
> > > Training could then be done via the ONNX Runtime Training API. Or even TorchSharp if that could load ONNX models.
> > > If I had time I would be working more on fulfilling this idea.
> > > Having both options would be best of course. Important thing is if ML.NET settles on using ONNX format, there are a lot of options. So happy with that too.
> > 
> > 
> > I support your idea, but most people will feel that it's too radical. And cutting the ecology of torch is a double-edged sword.
> 
> Cutting the C# /. NET ecosystem is even a more dangerous double-edged sword. Writing python and forcing C# developers to understand and use python.
> 
> They should take a look at what OpenCVSharp is doing. Beautiful apis that conforms to dotnet way of doing things

I get where you're coming from, but I'd urge you to consider the [simple reasoning](https://github.com/dotnet/TorchSharp#torchsharp) for the design decision. Put simply, given the enormous amount of existing documentation, tutorial and guide content and other collateral already created for PyTorch, keeping the naming and API conventions identical makes all this applicable to the .NET bindings as well.

No doubt there'll be an opportunity down the road to make a .NET-first API on top of the work done here, but in the interest of not starting entirely from scratch, this approach makes sense to me.

**Comment by Regenhardt** [comment]
Honestly I think this might not be a great reason. 

Assuming there will be many people actually porting python code over to ML.NET, after that is done everyone will have to continue using python naming conventions for no reason. 

With new people coming to machine learning every day though, I think many people will be confused by the different naming, which makes it look like an interop wrapper around a python library, which it is not. It's its own thing. 

I'm currently using python for my ML projects, but I've been trying to get into ML.NET hoping for some performance boost but also an easy way to build different frontends quickly, but most of all the static typing. Since we can build dotnet apps with almost no boilerplate and can even write dotnet interactive notebooks now, the IMO few advantages of python rapidly dwindle. 

I'd immensely prefer the naming also being in dotnet convention, matching the rest of the framework.

**Comment by GeorgeS2019** [comment]
>## Out of scope
> For this plan, it is currently out of scope to port over the PyTorch domain libraries (**Torch Vision**, **Torch Audio**, and **Torch Text**).

@briacht @luisquintanilla @migueldeicaza These out-of-scope topics defined back in [Aug 30, 2021](https://github.com/dotnet/machinelearning/issues/5918#issue-983175573) are now being actively developed.
- [Torch Vision](https://github.com/dotnet/TorchSharp/pull/609)
- [Torch Audio](https://github.com/dotnet/TorchSharp/pull/611)
- Torch Text will be next :-)

**Comment by GeorgeS2019** [comment]
Now TorchSharp's TorchVision and TorchAudio have their own nugets

nuget [Torchsharp TorchVision](https://www.nuget.org/packages/TorchVision)

nuget [Torchsharp TorchAudio](https://www.nuget.org/packages/TorchAudio)

**Comment by GeorgeS2019** [comment]
## Sept 2023 Update for [Deep Reinforcement Learning](https://github.com/dotnet/machinelearning/issues/181#issuecomment-1740019150)
